FIX Feature Selectors fail to route metadata when inside a Pipeline

kschluns opened the pull request:
#### Reference Issues/PRs
Towards #30527


#### What does this implement/fix? Explain your changes.
See #30527 for more context on the root issue.

The problem appears to be stemming from the fact that the Pipeline's `get_metadata_routing` function checks if each step's transformer (`trans`) has a `fit_transform` function, and if it does, it adds the following router mappings ([source](https://github.com/scikit-learn/scikit-learn/blob/d666202a9/sklearn/pipeline.py#L1353-L1358)):
```python
            if hasattr(trans, "fit_transform"):
                (
                    method_mapping.add(caller="fit", callee="fit_transform")
                    .add(caller="fit_transform", callee="fit_transform")
                    .add(caller="fit_predict", callee="fit_transform")
                )
```

All four impacted Feature Selector classes have a `fit_transform` function, which therefore requires a `.add(caller="fit_transform", callee=<???>)` mapping to exist in the downstream feature selector's `get_metadata_routing` function. The lack of this mapping is preventing metadata from being routed when `feature_selector.fit_transform()` is called, whereas metadata is successfully routed when `feature_selector.fit()` is called. 

Here is the current code for the `SelectFromModel` class which demonstrates the missing mapping ([source](https://github.com/scikit-learn/scikit-learn/blob/6e9039160/sklearn/feature_selection/_from_model.py#L499-L519)):
```python
    def get_metadata_routing(self):
        router = MetadataRouter(owner=self.__class__.__name__).add(
            estimator=self.estimator,
            method_mapping=MethodMapping()
            .add(caller="partial_fit", callee="partial_fit")
            .add(caller="fit", callee="fit"),
        )
        return router
```

I have tested that the issue is fixed by adding `.add(caller="fit_transform", callee="fit_transform")` to the `method_mapping` definition. It turns out that it doesn't matter what value you put for `callee` for reasons explained [in this comment](https://github.com/scikit-learn/scikit-learn/pull/30529#discussion_r1896131605). The only caveat to this is that the `callee` value has to be [a method that supports metadata routing](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/_metadata_requests.py#L90-L116).

#### Task List (from the [Pull Request Checklist](https://scikit-learn.org/stable/developers/contributing.html#pull-request-checklist))
- [x] Give your pull request a helpful title
- [x] Make sure your code passes the tests
- [x] Make sure your code is properly commented and documented, and make sure the documentation renders properly
- [x] Add non-regression tests specific to the issue and the bug fix
- [x] Add a changelog entry describing your PR changes (if necessary)

kschluns committed:
fixes metadata routing for feature selector fit_transform

glemaitre commented on sklearn/feature_selection/_from_model.py:
```diff
@@ -494,6 +494,7 @@ def get_metadata_routing(self):
         router = MetadataRouter(owner=self.__class__.__name__).add(
             estimator=self.estimator,
             method_mapping=MethodMapping()
+            .add(caller="fit_transform", callee="fit_transform")
```
```suggestion
            .add(caller="fit_transform", callee="fit")
```

On the top of the head, I would expect `callee` to be `fit` because this is the step that requires metadata for this specific selector.

However, I'm always confused with the routing/caller/callee :).

@StefanieSenger @adrinjalali would you mind to have a look.

adrinjalali replied:
we shouldn't have to add these _composite_ methods here, they should be auto generated from the simple methods. This might be an issue in the `_metadata_requests.py` that I need to debug to see where the issue is.

kschluns replied:
I was also curious about @glemaitre's question. It actually doesn't matter whether you use `callee="fit_transform"` or `callee="fit"` for the fix. Both values will result in successful metadata routing. 

I _think_ this is because the Feature Selector's `fit_transform` function passes the `**fit_params` to `self.fit()` without explicitly using the metadata routing functionality. Code snippet below ([source](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/base.py#L863)) shows that its only using the metadata routing for validation and has no impact on the `**fit_params`. 
```python
    def fit_transform(self, X, y=None, **fit_params):
        if _routing_enabled():
            transform_params = self.get_metadata_routing().consumes(
                method="transform", params=fit_params.keys()
            )
            if transform_params:
                warnings.warn(
                    (
                        f"This object ({self.__class__.__name__}) has a `transform`"
                        " method which consumes metadata, but `fit_transform` does not"
                        " forward metadata to `transform`. Please implement a custom"
                        " `fit_transform` method to forward metadata to `transform` as"
                        " well. Alternatively, you can explicitly do"
                        " `set_transform_request`and set all values to `False` to"
                        " disable metadata routed to `transform`, if that's an option."
                    ),
                    UserWarning,
                )
        if y is None:
            # fit method of arity 1 (unsupervised transformation)
            return self.fit(X, **fit_params).transform(X)
        else:
            # fit method of arity 2 (supervised transformation)
            return self.fit(X, y, **fit_params).transform(X)
```

To illustrate the function calls, in order:

1. `pipeline.fit()`:
    - First calls `process_routing(_method='fit')` to request the metadata to pass forward
    - Then calls `feature_selector.fit_transform()` and includes the routed metadata
2. `feature_selector.fit_transform()`:
    - does NOT call `process_routing()` 
    - calls `self.fit(X, y, **fit_params)`
3. `feature_selector.fit()`:
    - First calls `process_routing(_method='fit')` to request the metadata to pass forward
    - Then calls `estimator.fit()` and includes the routed metadata

Because step 2 doesn't participate in the metadata routing, it doesn't matter what the `callee` value is for the `feature_selector`. It just has to be one of the generically valid values defined by the class (e.g., `fit`, `transform`, `fit_transform`, etc).

glemaitre reviewed (commented):
So we should change also `sklearn/tests/test_metaestimators_metadata_routing/py` file. I assume that we forgot a specific method there for those selectors.

@StefanieSenger @adrinjalali Is there in the tests something where we nest those meta-estimator inside another meta-estimator, e.g. put them inside a `Pipeline` and make sure that the metadata are routed.

StefanieSenger commented:
> @StefanieSenger @adrinjalali Is there in the tests something where we nest those meta-estimator inside another meta-estimator, e.g. put them inside a Pipeline and make sure that the metadata are routed.

No, I don't think that we have those tests that go over more than two layers, unfortunately.

adrinjalali commented:
So our `MetadataRequest` object (used by non-routing estimators) does an automatic creation of _composite_ methods (such as `fit_transform`). However, `MetadataRouter` (used by routing estimators) does not.

`Pipeline`, for instance, has this code:

```py
            if hasattr(trans, "fit_transform"):
                (
                    method_mapping.add(caller="fit", callee="fit_transform")
                    .add(caller="fit_transform", callee="fit_transform")
                    .add(caller="fit_predict", callee="fit_transform")
                )
            else:
                (
                    method_mapping.add(caller="fit", callee="fit")
                    .add(caller="fit", callee="transform")
                    .add(caller="fit_transform", callee="fit")
                    .add(caller="fit_transform", callee="transform")
                    .add(caller="fit_predict", callee="fit")
                    .add(caller="fit_predict", callee="transform")
                )
```

And looking back at it, I'm wondering if `MetadataRouter` should also create the _composite_ requests from existing data the same way that `MetadataRequest` does.

kschluns commented:
> And looking back at it, I'm wondering if `MetadataRouter` should also create the _composite_ requests from existing data the same way that `MetadataRequest` does.

Do I understand right that both `Pipeline` and `SelectFromModel` are routing estimators? Therefore it doesn't take away from the fact that `SelectFromModel` still needs to supply a `.add(caller="fit_transform", callee=<???>)` in it's `get_metadata_routing()` function right? (i.e. regardless if it's auto-generated or not).

I ask because I'm wondering if we can allow this PR to go through with the short-term manual fix and then leave it as a future enhancement to automate the composite routing in the `MetadataRouter` class?

adrinjalali commented:
> I ask because I'm wondering if we can allow this PR to go through with the short-term manual fix and then leave it as a future enhancement to automate the composite routing in the MetadataRouter class?

That's true. I wouldn't mind that. This would need tests though, and ideally in the metadata routing common tests.

kschluns commented:
> That's true. I wouldn't mind that. This would need tests though, and ideally in the metadata routing common tests.

Sounds good! I can help with that. I'm new to contributing into sklearn though, so I may be in a bit over my head. But I'd like to give it a try first if that's okay. Is this the only file that we would need to add the tests to? --> `sklearn/tests/test_metaestimators_metadata_routing.py`

adrinjalali commented:
Yep, that's the file. And thanks for contributing 😊

kschluns committed:
added non-regression tests for feature_selector fix

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
linter fix

kschluns commented:
Hey @adrinjalali I just added a non-regression test. Could you please review and let me know if it's satisfactory? 

Also I have a couple questions about the two remaining tasks I have:
- [ ] Make sure your code passes the tests 
_**Do I need manually run some other test suites to confirm no issues or are all tests run as part of the CI? (e.g. tests other than `sklearn/tests/test_metaestimators_metadata_routing.py`)**_ 
- [ ] Add a changelog entry describing your PR changes (if necessary)
_**I see that the Check Changelog is failing now that I've added a non-regression test. Does this mean I need to add a changelog entry now?**_

kschluns committed:
simplified feature_selector metadata routing test

kschluns committed:
Add change log for PR 30529

kschluns commented:
Hey @adrinjalali I just finished the remaining tasks.  Could you please review and let me know if this is satisfactory? 
- [x] Make sure your code passes the tests 
_**Ran the following tests. Let me know if I should run any other tests.**_
`pytest sklearn/feature_selection`
`pytest sklearn/tests/test_metadata_routing.py`
`pytest sklearn/tests/test_metaestimators_metadata_routing.py`
`pytest sklearn/tests/test_metaestimators.py`
`pytest doc/modules/feature_selection.rst`
- [x] Add non-regression tests specific to the issue and the bug fix
_**Let me know if the new `test_feature_selectors_in_pipeline` test is sufficent!**_
- [x] Add a changelog entry describing your PR changes (if necessary)

kschluns committed:
fix test coverage issue

kschluns committed:
flake8 fix

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
Update doc/whats_new/upcoming_changes/sklearn.feature_selection/30529.fix.rst

Co-authored-by: Adrin Jalali <adrin.jalali@gmail.com>

kschluns committed:
Added metadata routing via fit_transform for IterativeImputer

kschluns committed:
Improved test_metaestimators_in_pipeline

kschluns commented on sklearn/tests/test_metaestimators_metadata_routing.py:
```diff
@@ -697,6 +703,87 @@ def test_error_on_missing_requests_for_sub_estimator(metaestimator):
+    if ("estimator" not in metaestimator) or (
+        metaestimator_class.__module__.split(".")[1]
+    ) != "feature_selection":
+        return
```
Latest set of commits have addressed this. Resolving comment.

kschluns commented on sklearn/tests/test_metaestimators_metadata_routing.py:
```diff
@@ -697,6 +703,87 @@ def test_error_on_missing_requests_for_sub_estimator(metaestimator):
+        assert (
+            selected_features_with_sample_weights
+            != selected_features_without_sample_weights
+        )
```
Agreed! Latest set of commits have addressed this. Resolving comment.

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
Update change log 30529.fix.rst

kschluns committed:
removing unused variables

kschluns committed:
ruff linter fix

kschluns commented:
@adrinjalali thank you for your review! I agreed with all the suggestions and made improvements to the tests accordingly. Can you please review again? 

Note that in the process of making these changes, I realized that the prior implementation of the `check_recorded_metadata` function is resulting in many tests passing trivially, particularly in `test_setting_request_on_sub_estimator_removes_error`. Please see the note left in the `check_recorded_metadata` docstring for an explanation. 

I added functionality to `check_recorded_metadata` that allows the `parent` parameter to be `None`, which successfully works around the issue, but is only valid if the test doesn't care about the value of `parent` (which is true for the majority of tests, but not all). I tested setting `parent=None` for all tests and there's only three failures to deal with, but I don't want to increase the scope of this PR beyond what's intended. Therefore, I propose only setting `parent=none` for `test_metaestimators_in_pipeline` and making a future PR to fix the issue of trivial passed tests in the other test sets. 

What do you think?

kschluns committed:
fixed test coverage issue

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns commented:
Hey @adrinjalali, are you able to review this PR sometime this week?

adrinjalali commented:
@kschluns I'll have a look this week.

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns commented:
Hey @adrinjalali any update here? Is it normal for a PR to take this long to get approved or is there something about the proposed changes here that is making it difficult to review and approve? If you're just too busy, can we see if someone else is able to review this PR?

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing

kschluns commented:
Hello? It's me.

kschluns commented:
Hey @adrinjalali! Can you please take a look at this PR?

adrinjalali commented on sklearn/tests/metadata_routing_common.py:
```diff
@@ -57,20 +57,41 @@ def check_recorded_metadata(obj, method, parent, split_params=tuple(), **kwargs)
+        The parent method which should have called `method`, or otherwise in
+        the context of metadata routing referred to as 'caller'.
+
+        Sub-estimator metadata is only checked if the `caller` method matches
+        the value defined by `parent`. If `parent` is None, the target
+        sub-estimator metadata is checked regardless of the `caller` method.
+
+        NOTE: many metaestimators call the subestimator in roundabout ways
+        and this makes it very difficult to know what method name to use for
+        `parent`. If misspecified, it results in tests passing trivially. For
+        example, when fitting the RFE metaestimator, RFE.fit() calls RFE._fit()
+        , which then calls subestimator.fit(). In this case, the user
+        configuring the test should set method="fit" and parent="_fit",
+        otherwise the test will pass trivially.
```
I'm sure you've done a great job going through this bit of code figuring it out, but from the change in the docstring, it's not clear to me exactly what the issue is. Could you please add a test specifically for the change here, to make it clear for future developers?

kschluns replied:
Should I actually add a test in the `sklearn/tests/test_metaestimators_metadata_routing.py`, mark it with `@pytest.mark.skip` since it is expected to fail, then just add a note in the docstring referencing the test?

adrinjalali replied:
Since I don't really understand this comment, seeing that test helps.

kschluns replied:
My question was more around the logistics of adding a test that I know is going to fail, without it messing up any CI pipelines. For example, at work we enforce 100% of our tests to pass in order for PRs to get merged, so I just haven't encountered this situation before. Wasn't sure if that would be a problem here and if it is, is `@pytest.mark.skip` the right way to handle it?

adrinjalali replied:
We won't be merging w/o CI being green. Once you write a test, I'd know what needs to be done to fix it. Here I just don't understand what the comment is saying.

kschluns replied:
got it, will write up a test shortly

adrinjalali commented on sklearn/tests/test_metaestimators_metadata_routing.py:
```diff
@@ -697,6 +701,91 @@ def test_error_on_missing_requests_for_sub_estimator(metaestimator):
+        metaestimator_instance = metaestimator_class(**kwargs)
+        extra_method_args = metaestimator.get("method_args", {}).get(method_name, {})
+
+        if not hasattr(metaestimator_instance, "transform"):
+            # This test only makes sense for metaestimators that can be
+            # intermediate steps in a Pipeline (e.g. transform method exists)
+            return
```
this check can go up, in the form of `if not hasattr(metaestimator_class, "transform")`

kschluns replied:
_**TL;DR - The suggested change causes four new tests to fail, which were being skipped prior (which was intended). Can I leave the code the way it was?**_

The four new test failures in `test_metaestimators_in_pipeline`: 
- GridSearchCV
- RandomizedSearchCV
- HalvingGridSearchCV
- HalvingRandomSearchCV

The test failure message is below:
```python
TypeError: All intermediate steps should be transformers and implement fit and transform or be the string 'passthrough'
```

For these classes, `hasattr(metaestimator_class, "transform") --> True`, but `hasattr(metaestimator_instance, "transform") --> False`. 

The difference in the class check vs the instance check is due to the fact that these four classes have the following decorator, which delays defining the transform function until after the class has been instantiated.
```python
@available_if(_search_estimator_has("transform"))
def transform(self, X):
```

These classes are being instantiated with a ConsumingClassifier estimator, which doesn't have a transform method and that's what's causing the difference.

adrinjalali reviewed (commented):
Thanks a lot @kschluns , and sorry for the late review. This is looking much better now.

kschluns committed:
Merge branch 'main' into fix_feature_selector_metadata_routing
